File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-2.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
 
public class Solution {
    public ListNode MergeTwoLists(ListNode list1, ListNode list2) {
        if (list1 == null && list2 == null) {
            return list1;
        }
        if (list2 == null) {
            return list1;
        }
        if (list1 == null) {
            return list2;
        }
        ListNode head;

        if (list1.val <= list2.val) {
            head = list1;
            helper(list1, list1.next, list2);
        }
        else {
            head = list2;
            helper(list2, list1, list2.next);
        }
        return head;
    }

    public void helper(ListNode curEnd, ListNode list1, ListNode list2) {
        if (list1 == null && list2 == null) {
            return;
        }
        if (list1 == null) {
            curEnd.next = list2;
            return;
        }
        if (list2 == null) {
            curEnd.next = list1;
            return;
        }
        if (list1.val <= list2.val) {
            curEnd.next = list1;
            helper(list1, list1.next, list2);
        }
        else {
            curEnd.next = list2;
            helper(list2, list1, list2.next);
        }
    }
}

File: Data Structures & Algorithms/reorder-linked-list/submission-2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        front = head;
        mid = head;
        end = head;
        while end.next is not None:
            mid = mid.next
            end = end.next
            if end.next is not None:
                end = end.next
        end = mid.next
        mid.next = None
        end = self.reverse(end)
        cur = head
        front = front.next
        while end:

            cur.next = end
            end = end.next
            cur = cur.next

            cur.next = front
            front = front.next
            cur = cur.next

        if front:
            cur.next = front
            cur = cur.next
            
        if cur:
            cur.next = None
        
        


    def reverse(self, head) -> ListNode:
        prev = None
        cur = head
        while cur:
            nxt = cur.next
            cur.next = prev
            prev = cur
            cur = nxt
        return prev



